packages/components/src/Menu/Menu.tsx

```diff
@@ -277,10 +294,87 @@ export function Menu({
   }
 }
 
-interface SectionHeaderProps {
-  readonly text: string;
-  readonly UNSAFE_style?: CSSProperties;
-  readonly UNSAFE_className?: string;
+type AnimationState = "unmounted" | "hidden" | "visible";
+interface MenuAnimationContextValue {
+  state: AnimationState;
+  setState: React.Dispatch<React.SetStateAction<AnimationState>>;
+}
+const MenuAnimationContext = createContext<MenuAnimationContextValue | null>(
+  null,
+);
+
+function useMenuAnimation(): MenuAnimationContextValue {
+  const ctx = useContext(MenuAnimationContext);
+
+  if (!ctx) {
+    throw new Error("MenuAnimationContext used outside provider");
+  }
+
+  return ctx;
+}
+
+function MenuComposable({ children, onOpenChange }: MenuComposableProps) {
+  // Use positional arguments to determine the trigger and content
+  // Avoids parsing/iterating over the children
+  const [trigger, menu] = React.Children.toArray(children);
+
+  const [animation, setAnimation] = useState<AnimationState>("unmounted");
+
+  return (
+    <div className={styles.wrapper}>
+      <MenuAnimationContext.Provider
+        value={{ state: animation, setState: setAnimation }}
+      >
+        <AriaMenuTrigger
+          onOpenChange={isOpen => {
+            setAnimation(isOpen ? "visible" : "hidden");
+            onOpenChange?.(isOpen);
+          }}
+        >
+          {trigger}
+          {/* Keep Popover mounted while exiting, but do not animate it. */}
+          <AriaPopover isExiting={animation === "hidden"}>
+            {({ placement }) => {
+              if (React.isValidElement(menu)) {
+                return React.cloneElement(
+                  menu as ReactElement<MenuContentComposableProps>,
+                  {
+                    placement,
+                  },
+                );
+              }
```

Comment:
I know we want to avoid cloning elements but there's no other way to do this as far as I can tell

the problem is that our animation needs to know where the popover was placed. ie. bottom or top

since our animation lives in JS via framer-motion, that information has to be in JS. the only place that exposes this detail is the render props for `Popover`

trying to set the placement as state gets messy because we're in the middle of rendering another component and it throws errors. I've tried that with the Provider approach and it wouldn't work.

I've seen some new additions to Popover that allows origin aware animations, though I believe that's in css - not available in JS/React.

https://react-spectrum.adobe.com/releases/2025-08-25.html